### ISSUE TITLE
Impact of training batch size for in-batch softmax loss

### ISSUE THREAD

**Issue by esriis** [tensorflow/recommenders#394]
Hi, I was wondering if there are any guidelines/references/intuition on choosing the batch size for the training data when using in-batch softmax loss?

In the paper **Self-supervised Learning for Large-scale Item Recommendations**, it states "Note that the training batch size in batch softmax is critical for model quality as it determines the number of negatives used for each positive item." And from the Tensorflow-Recommenders tutorials and some of the related papers, it seems that batch sizes in the range 1,000 - 10,000 are the most common, so I'm curious about e.g. the expected effect of using smaller batches. Or if there is some intuition about this topic for someone with a background in statistics?

**Comment by maciejkula** [comment]
For in-batch softmax, there are two effects of increasing the batch size:

- You have more in-batch negatives. If you have strong reasons to believe the positive item should be recommended, and no other items should, this will give you a better model. On the other hand, other items might also be good choices; treating them as negatives here might make the model worse. It's likely that there is a happy middle here.
- Using larger batches affects model training dynamics, often negatively (see [this paper](https://openreview.net/pdf?id=H1oyRlYgg) for some discussion).

So making your batch size smaller could work either way - as always, this is a hyperparameter best tuned specifically for your use case.
